### PR TITLE
Port FlipSign to Operator2

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
`FlipSign` is a template in PennyLane

**Description of the Change:**
The template is ported to `Operator2`.

Breaking changes:
- The basis state for which the sign is flipped is now homogeneously called `state` everywhere, instead of `n` in the inputs and `arr_bin` in the `hyperparameters`.
- Input wires are handled like for other PennyLane operators rather than with the custom logic present before. For example `""` is a valid wire label in PennyLane, so it is a valid input argument for `wires` in `FlipSign`, which was not the case before.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-126040]

### CODE REVIEW

File: pennylane/templates/subroutines/flip_sign.py [hunk 1]
```diff
@@ -64,118 +70,93 @@ def circuit():
...
-        self._hyperparameters = {"arr_bin": n}
-        super().__init__(wires=wires)
-
-    @property
-    def resource_params(self):
-        return {
-            "num_wires": len(self.wires),
-            "arr_bin": self.hyperparameters["arr_bin"],
-        }
+    wires_argnames = ("wires",)
```
Comment (andrijapau) [inline pennylane/templates/subroutines/flip_sign.py]: `"wires"` is the assumed argname by default 🙂 
```suggestion
```

File: pennylane/templates/subroutines/flip_sign.py [hunk 2]
```diff
@@ -64,118 +70,93 @@ def circuit():
...
+    compilable_argnames = ("state",)
+    arg_specs = {"wires": Wire[-1]}
```
Comment (andrijapau) [inline pennylane/templates/subroutines/flip_sign.py:70]: So the "canonical" `state` is a sequence of integers, right?
```suggestion
    arg_specs = {"state": Int[-1], "wires": Wire[-1]}
```

File: pennylane/templates/subroutines/flip_sign.py
Comment (andrijapau) [inline pennylane/templates/subroutines/flip_sign.py:1]: Totally optional, but I recommend running `pylint --enable useless-suppression` whenever possible. I'm sure there are a bunch of stale pylint disables now. 😄

**Review by andrijapau — COMMENTED** [review]
Looks really nice! Just some comments. 🎉

File: pennylane/templates/subroutines/flip_sign.py [hunk 3]
```diff
@@ -64,118 +70,93 @@ def circuit():
...
+                )
+            state = tuple(map(int, math.int_to_binary(state, num_wires)))
+        else:
+            if num_wires != len(state):
+                raise ValueError(
+                    "The basis state and wires must have equal length, "
+                    f"but got {len(state)} and {num_wires}."
+                )
+            state = tuple(state)
+        return state
```
Comment (andrijapau) [inline pennylane/templates/subroutines/flip_sign.py]: Bit easier to read IMO
```suggestion
        if isinstance(state, int):
            if not 0 <= state < 2**num_wires:
                raise ValueError(
                    "The given basis state must be a non-negative integer smaller "
                    f"than {2**num_wires}, but got {state}."
                )
            return tuple(map(int, math.int_to_binary(state, num_wires)))
            
        if num_wires != len(state):
            raise ValueError(
                "The basis state and wires must have equal length, "
                f"but got {len(state)} and {num_wires}."
            )
        return tuple(state)
```

File: pennylane/templates/subroutines/flip_sign.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/templates/subroutines/flip_sign.py]: Ah, I copied this from `BasisRotation` 🙃

File: pennylane/templates/subroutines/flip_sign.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/templates/subroutines/flip_sign.py:70]: Yes, but it is not an array like,but a tuple. Is that still correctly described with `Int[-1]`?

File: pennylane/templates/subroutines/flip_sign.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/templates/subroutines/flip_sign.py]: Ah yes, nice! Pulling this out into its own method was the last thing I did, so I didn't realize this potential :D

File: pennylane/templates/subroutines/flip_sign.py
Comment (dwierichs) [reply pennylane/templates/subroutines/flip_sign.py:1]:
```
--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```
👀

Comment (andrijapau) [reply pennylane/templates/subroutines/flip_sign.py:1]: 😲

File: pennylane/templates/subroutines/flip_sign.py [hunk 4]
```diff
@@ -64,118 +70,93 @@ def circuit():
...
+        state = self._canonicalize_state(state, len(wires))
+        super().__abstract_init__(state, wires)
 
     @property
     def num_params(self):
         return 0
 
     @staticmethod
-    def compute_decomposition(wires, arr_bin):  # pylint: disable=arguments-differ
+    def compute_decomposition(state, wires):  # pylint: disable=arguments-differ
```
Comment (andrijapau) [inline pennylane/templates/subroutines/flip_sign.py]: This method shouldn't be needed anymore 😃

File: pennylane/templates/subroutines/flip_sign.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/templates/subroutines/flip_sign.py]: Why was it in `BasisRotation`?

Comment (andrijapau) [reply pennylane/templates/subroutines/flip_sign.py]: Probably slipped through the cracks when it was ported

File: pennylane/templates/subroutines/flip_sign.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/templates/subroutines/flip_sign.py:70]: `Int[-1]` would be a dynamic array of ones. Do compilable argnames go into argspecs at all? Looks like @mudit2812 's keeping it out of `PauliRot`: [pennylane/ops/qubit/parametric_ops_multi_qubit.py](https://github.com/PennyLaneAI/pennylane/pull/9897/changes#diff-d01a7187bea6fdfd949bbc9eede362c1a14b393beed01f668b5885ab82f9c635)

File: pennylane/templates/subroutines/flip_sign.py [hunk 5]
```diff
@@ -64,118 +70,93 @@ def circuit():
...
+            raise ValueError("At least one wire is required.")
+        state = self._canonicalize_state(state, num_wires)
+        super().__init__(state, wires)
+
+    def __abstract_init__(self, state, wires):  # pylint: disable=arguments-differ
+        state = self._canonicalize_state(state, len(wires))
+        super().__abstract_init__(state, wires)
 
     @property
     def num_params(self):
```
Comment (andrijapau) [inline pennylane/templates/subroutines/flip_sign.py]: This could probably be deleted too as it gets the `num_params` from the dynamic args (of which there are none in this template).

File: pennylane/templates/subroutines/flip_sign.py [hunk 6]
```diff
@@ -33,17 +37,19 @@ class FlipSign(Operation):
...
 
     **Example**
 
-    This template changes the sign of the basis state passed as an argument.
-    In this example, when passing the element ``[1, 0]``, we will change the sign of the state :math:`|10\rangle`.
-    We could alternatively pass the integer ``2`` and get the same result since its binary representation is ``[1, 0]``.
+    This template changes the sign of the basis state passed as an argument. In this example,
+    when passing the element ``[1, 0]``, we will change the sign of the state :math:`|10\rangle`
+    on two qubits. We could alternatively pass the integer ``2`` and get the same result since
+    its binary representation on two bits is ``[1, 0]``.
```
Comment (andrijapau) [inline pennylane/templates/subroutines/flip_sign.py]:
```suggestion
    its two-bit binary representation is ``(1, 0)``.
```

File: pennylane/templates/subroutines/flip_sign.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/templates/subroutines/flip_sign.py]: I'm removing it from `BasisRotation`.

File: pennylane/templates/subroutines/flip_sign.py [hunk 7]
```diff
@@ -33,17 +37,19 @@ class FlipSign(Operation):
 
     FlipSign(n) :math:`|m\rangle = |m\rangle` if :math:`m \not = n`,
 
-    where n is the basis state to flip and m is the input.
+    where :math:`n` is the basis state (argument ``state``) to flip and :math:`m` is the input.
 
     Args:
-        n (array[int] or int): binary array or integer value representing the state on which to flip the sign
+        state (Sequence[int] or array[int] or int): binary array or integer value representing
+            the state on which to flip the sign
```
Comment (andrijapau) [inline pennylane/templates/subroutines/flip_sign.py]: 💥 Extreme nitpick 💥 
```suggestion
        state (Sequence[int] or array[int] or int): integer or binary sequence representing the basis state whose sign is to be flipped
```

File: pennylane/templates/subroutines/flip_sign.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/templates/subroutines/flip_sign.py:70]: The guide also indicates this, but doesn't make an ultimate call.

File: pennylane/templates/subroutines/flip_sign.py (hunk 5 above)
Comment (dwierichs) [reply pennylane/templates/subroutines/flip_sign.py]: Again, I copied `BasisRotation` behaviour. Will delete it there as well.

File: pennylane/templates/subroutines/flip_sign.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/templates/subroutines/flip_sign.py:70]: In any case, it seems _okay_ to not include it. :)

Comment (mudit2812) [reply pennylane/templates/subroutines/flip_sign.py:70]: Thanks for pointing this out. `arg_specs` is to declare _types_, which only works for `dynamic_argnames` and non-hybrid `wire_argnames`. Static values (and the pytree structures of hybrid arguments) are defined by concrete values/structures, so they don't go into `arg_specs`. I'll update the migration guide to say the same.

File: pennylane/templates/subroutines/flip_sign.py (hunk 4 above)
Comment (albi3ro) [inline pennylane/templates/subroutines/flip_sign.py]: Do we need to keep `compute_decomposition` around? In #9910 I'm moving the graph decomposition fallback from `Operator2.decomposition` to `Operator2.compute_decomposition`, so that should help us keep around any unit tests that change.

File: pennylane/templates/subroutines/flip_sign.py (hunk 2 above)
Comment (andrijapau) [reply pennylane/templates/subroutines/flip_sign.py:70]: Yeah, on second thought the above makes sense. We should update the docs though for sure!

File: pennylane/templates/subroutines/flip_sign.py (hunk 4 above)
Comment (andrijapau) [reply pennylane/templates/subroutines/flip_sign.py]: Was removed here https://github.com/PennyLaneAI/pennylane/pull/9924#discussion_r3674742573!

File: pennylane/templates/subroutines/flip_sign.py [hunk 8]
```diff
@@ -33,17 +37,19 @@ class FlipSign(Operation):
 
     FlipSign(n) :math:`|m\rangle = |m\rangle` if :math:`m \not = n`,
 
-    where n is the basis state to flip and m is the input.
+    where :math:`n` is the basis state (argument ``state``) to flip and :math:`m` is the input.
 
     Args:
-        n (array[int] or int): binary array or integer value representing the state on which to flip the sign
+        state (Sequence[int] or array[int] or int): integer or binary sequence representing
```
Comment (albi3ro) [inline pennylane/templates/subroutines/flip_sign.py]: Are we still supporting `array[int]`? I thought we were outlawing that to make it compilable instead of dynamic? Might even be worth being a bit more specific for `tuple[int], list[int], int` so we don't imply that it can be an array.

**Review by andrijapau — APPROVED** [review]
LGTM 🚀

File: tests/capture/test_templates.py [hunk 9]
```diff
@@ -1768,6 +1767,27 @@ def qfunc():
...
+        jaxpr = jax.make_jaxpr(qfunc)(wires)
+
+        assert len(jaxpr.eqns) == 1
+
+        eqn = jaxpr.eqns[0]
+        assert_eqn_matches_op(eqn, qp.FlipSign)
+        assert eqn.invars == jaxpr.jaxpr.invars
+        assert len(eqn.outvars) == 1
+        assert isinstance(eqn.outvars[0], jax.core.DropVar)
+
```
Comment (JerryChen97) [inline tests/capture/test_templates.py]: Could we also explicitly test the survival of `state`

File: tests/templates/subroutines/test_flip_sign.py
Comment (JerryChen97) [inline tests/templates/subroutines/test_flip_sign.py:1]: PR description said that `""` is a valid input for `wires` now. Is it still the case? Or maybe I missed some condition for that to be true

**Review by JerryChen97 — COMMENTED** [review]
Overall LGTM! Just some questions

**Review by JerryChen97 — APPROVED** [review]

File: tests/templates/subroutines/test_flip_sign.py
Comment (KetpuntoG) [reply tests/templates/subroutines/test_flip_sign.py:1]: It seems to still be the case
I added a test to check that indeed it is working

Screenshot 2026-08-06 at 9 35 21 AM
